README.md
# rust-wurfl
rust bindings to InFuze C API (libwurfl).

### wurfl-sys

This directory contains the bindings to the InFuze libwurfl (which MUST be installed to make everything work).
It requires Rust edition 2018 to work.

### rust-wurfl
rust-wurfl is the root project that contains the `wurfl-sys`. 
It provides an abstraction layer built on top of the bindings of the wurfl-sys crate. 
It exposes two structs Wurfl and Device with `lookup*` methods by User-Agent, Headers map and `device_id`.
Wurfl struct also exposes methods to update WURFL.xml (the so called WURFL-updater) like `updater_runonce`, `updater_start`, `updater_stop`.
Device struct exposes `get_capability`, `get_virtual_capability` and aggregate methods to get groups of them.
Wurfl and Device structs implement `Drop` trait to deallocate their resources.

To compile and run the rust-wurfl example you just need to run `cargo run --example example` from the rust-wurfl directory.
It will print something like this: 
```
Starting WURFL wrapper usage sample!
WURFL API Version created: 1.12.1.0
Last load time:  Thu Jun 24 12:38:03 2021
WURFL info:  Root:/usr/share/wurfl/wurfl.zip:for WURFL API 1.12.1.0 - min, db.scientiamobile.com - 2021-04-27 12:18:27:2021-04-27 12:27:39 -0400
Printing capability name from Iterator
> is_wireless_device <
Printing all capability names --------------------------
< is_wireless_device >
< model_name >
< ux_full_desktop >
< pointing_method >
< physical_screen_width >
< preferred_markup >
< is_tablet >
< resolution_height >
< resolution_width >
< can_assign_phone_number >
< marketing_name >
< brand_name >
< physical_screen_height >
< mobile_browser >
< device_os_version >
< is_smarttv >
< xhtml_support_level >
< mobile_browser_version >
< device_os >
---------------------------------------------------------
Printing all virtual capability names -------------------
< advertised_browser >
< advertised_browser_version >
< advertised_device_os >
< advertised_device_os_version >
< complete_device_name >
< device_name >
< form_factor >
< is_android >
< is_app >
< is_app_webview >
< is_full_desktop >
< is_html_preferred >
< is_ios >
< is_largescreen >
< is_mobile >
< is_phone >
< is_robot >
< is_smartphone >
< is_touchscreen >
< is_windows_phone >
< is_wml_preferred >
< is_xhtmlmp_preferred >
< advertised_app_name >
< generalized_browser_type >
< generalized_os_brand >
< pixel_density >
< is_generic >
// more output follows...
```

Please note that the number and availability of capabilities may change depending on yout WURFL license.


CHANGELOG.md
## rust-wurfl [0.8.0] - InFuze 1.12.2.0 - 2021-07-20

Initial release
- Wurfl provides detection methods:
  - lookup_useragent
  - lookup_with_headers
  - lookup_device_id_with_headers
  - lookup_device_id
    
- Device methods:
  - get_capability
  - get_virtual_capability
  - get_capabilities
  - get_virtual_capabilities
  - has_capability
  - has_virtual_capability

 - Wurfl Updater methods
   - updater_runonce
   - updater_start
   - updater_stop
   - other updater configuration methods

src/device.rs
/// Device provides access to device static and virtual capabilities
pub struct Device {
    // we use this attribute to silence the "variable is never read" warning
    #[allow(dead_code)]
    wurfl: wurfl_handle,
    device: wurfl_device_handle,
}

/// Device provides access to device static and virtual capabilities and other device data
impl Device {
    /// returns the WURFL device unique ID
    pub fn get_device_id(&self) -> &str {
        let d_id = unsafe { wurfl_device_get_id(self.device) };
        return to_str(d_id);
    }

    /// Returns the unique ID of this device ancestor in the WURFL hierarchy
    pub fn get_root_ID(&self) -> &str {
        let d_root_id = unsafe { wurfl_device_get_root_id(self.device) };
        return to_str(d_root_id);
    }

    /// Returns true if this device is a root device in the WURFL hierarchy
    pub fn is_root(&self) -> bool {
        let is_root = unsafe { wurfl_device_is_actual_device_root(self.device) };
        if is_root == 0 {
            return false;
        }
        return true;
    }

    /// Returns the original UserAgent of matched device (the one passed to lookup)
    pub fn get_original_user_agent(&self) -> &str {
        let d_orig_ua = unsafe { wurfl_device_get_original_useragent(self.device) };
        return to_str(d_orig_ua);
    }

    /// Return a normalized version of this device User-Agent
    pub fn get_normalized_user_agent(&self) -> &str {
        let d_norm_ua = unsafe { wurfl_device_get_normalized_useragent(self.device) };
        return to_str(d_norm_ua);
    }

    /// Returns default UserAgent of matched device (might be different from UA passed to lookup)
    pub fn get_user_agent(&self) -> &str {
        let d_ua = unsafe { wurfl_device_get_useragent(self.device) };
        return to_str(d_ua);
    }

    /// Returns the device value of the capability with the given name
    pub fn get_capability(&self, capability_name: &str) -> Option<&str> {
        let c_str_cap_name = match CString::new(capability_name) {
            Ok(cn) => cn,
            Err(_) => return None
        };
        let cap_value = unsafe { wurfl_device_get_capability(self.device, c_str_cap_name.as_ptr()) };
        if cap_value.is_null() {
            return None;
        }
        return Some(to_str(cap_value));
    }

    /// Returns the device value of the virtual capability with the given name
    pub fn get_virtual_capability(&self, virtual_capability_name: &str) -> Option<&str> {
        let c_str_vcap_name = match CString::new(virtual_capability_name) {
            Ok(cn) => cn,
            Err(_) => return None
        };
        let vcap_value = unsafe { wurfl_device_get_virtual_capability(self.device, c_str_vcap_name.as_ptr()) };
        if vcap_value.is_null() {
            return None;
        }
        return Some(to_str(vcap_value));
    }

    /// Returns a map -> key:cap_name,value:cap_value for the calling device
    pub fn get_capabilities(&self, cap_names: &[&str]) -> Result<HashMap<String, String>, WurflError> {
        let mut caps = HashMap::new();
        for cap_name in cap_names {
            let msg = format!("Unable to convert into a CString capability name  {}", cap_name);
            // this way string doesn't take ownership of cap_name
            let c_str_cap_name = match CString::new(cap_name.clone()) {
                Ok(cn) => cn,
                Err(_) => return Err(WurflError { msg: msg }),  // failed to convert to C string
            };

            let cap_ptr = unsafe { wurfl_device_get_capability(self.device, c_str_cap_name.as_ptr()) };
            if !cap_ptr.is_null() {
                let c_str = unsafe { CStr::from_ptr(cap_ptr) };
                // the string coming from get_capability should be UTF-8 compliant
                let cap_value = match c_str.to_str() {
                    Ok(cv) => cv.to_string(),
                    Err(utf_err) => {
                        let emsg = format!("Capability value for {} is an invalid UTF-8 string, it was valid until character {}", &cap_name, utf_err.valid_up_to());
                        return Err(WurflError { msg: emsg });
                    }
                };
                caps.insert(cap_name.to_string(), cap_value);
            }
        };
        return Ok(caps);
    }

    /// Returns a map -> key:vcap_name,value:vcap_value for the calling device
    pub fn get_virtual_capabilities(&self, vcap_names: &[&str]) -> Result<HashMap<String, String>, WurflError> {
        let mut vcaps = HashMap::new();
        for vcap_name in vcap_names {
            let msg = format!("Unable to convert  virtual capability name  {}! into a CString", vcap_name);
            // this way string doesn't take ownership of cap_name
            let c_str_vcap_name = match CString::new(vcap_name.clone()) {
                Ok(vcn) => vcn,
                Err(_) => return Err(WurflError { msg: msg }),  // failed to convert to C string
            };
            let vcap_ptr = unsafe { wurfl_device_get_virtual_capability(self.device, c_str_vcap_name.as_ptr()) };
            if !vcap_ptr.is_null() {
                let c_str = unsafe { CStr::from_ptr(vcap_ptr) };
                // the string coming from get_capability should be UTF-8 compliant
                let vcap_value = match c_str.to_str() {
                    Ok(cv) => cv.to_string(),
                    Err(utf_err) => {
                        let emsg = format!("Virtual capability value for {} is an invalid UTF-8 string, it was valid until characer {}", &vcap_name, utf_err.valid_up_to());
                        return Err(WurflError { msg: emsg });
                    }
                };
                vcaps.insert(vcap_name.to_string(), vcap_value);
            }
        };
        return Ok(vcaps);
    }

    /// Return the type of matching performed to detect this device
    pub fn get_match_type(&self) -> MatchType {
        let mt = unsafe { wurfl_device_get_match_type(self.device) };
        match mt {
            0 => MatchType::WurflMatchTypeExact,
            1 => MatchType::WurflMatchTypeConclusive,
            2 => MatchType::WurflMatchTypeRecovery,
            3 => MatchType::WurflMatchTypeCatchall,
            // Value 4 (performance) is deprecated and shouldn0t be returned anymore by C API.
            5 => MatchType::WurflMatchTypeNone,
            6 => MatchType::WurflMatchTypeCached,
            // Catch all match: any other value is mapped as Match None
            _ => MatchType::WurflMatchTypeNone
        }
    }
}

// Device memory deallocation
impl Drop for Device {
    fn drop(&mut self) {
        //println!("Destroying... WURFL device");
        unsafe { wurfl_device_destroy(self.device) };
    }
}

// makes device (and it internal objects) transferable between threads
unsafe impl Send for Device {}

// assumes Device struct thread safety, which, in our case is supported by the underlying Infuze C library
unsafe impl Sync for Device {}


examples/hyper_example.rs
use std::convert::Infallible;
use std::net::SocketAddr;
use std::sync::Arc;

use hyper::{Body, Request, Response, Server};
use hyper::service::{make_service_fn, service_fn};

use wurfl::{Wurfl, WurflCacheProvider};

const DEFAULT_WURFL_FILE_PATH: &str = "/usr/share/wurfl/wurfl.zip";

#[tokio::main]
pub async fn main() {

    // This example is a variation of the "getting started" server example of hyper framework,
    // with the addition of a function that performs a device detection and returns a string
    // containing the WURFL device ID and its brand, model and marketing name (if available).

    // by default, this example will use the default WURFL file path for Linux machines,
    // but tou can set a custom WURFL file path by setting the WURFL_FILE_PATH environment variable.

    // retrieve WURFL file path. First from a custom location
    let key = "WURFL_FILE_PATH";
    let wurfl_path = match std::env::var(key) {
        Ok(p) => p,
        _ => {
            println!("env var WURFL_FILE_PATH not set, using default value for WURFL file path");
            DEFAULT_WURFL_FILE_PATH.to_owned()
        }
    };

    // First, create a Wurfl engine instance with cache provider
    let wurfl_res = Wurfl::new(&*wurfl_path, None, None,
                               WurflCacheProvider::LRU, Some("100000"));
    let engine = match wurfl_res {
        Ok(engine) => engine,
        Err(error) => panic!("Problem initializing wurfl: {:?}", error),
    };
    println!("WURFL API Version created: {}", engine.get_api_version());
    // now make it thread safe
    let safe_engine = Arc::new(engine);

    // A `Service` is needed for every connection, so this
    // creates one wrapping our `detect` function.
    let make_svc = make_service_fn(move |_conn| {
        let safe_engine_clone = Arc::clone(&safe_engine);
        async {
            Ok::<_, Infallible>(service_fn(move |req| {
                let response = detect(req, &safe_engine_clone);
                async { Ok::<_, Infallible>(Response::new(Body::from(response))) }
            }))
        }
    });

    // We'll bind the server to 127.0.0.1:3000
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    let server = Server::bind(&addr).serve(make_svc);
    match server.await {
        Err(_) => panic!("An error occurred while running WURFL hyper server example, shutting down"),
        _ => (),
    }
}

// Actual device detection: returns a string with wurfl_id and virtual capability complete_device_name
fn detect(_req: Request<Body>, safe_engine: &Arc<Wurfl>) -> String {
    let device = match safe_engine.lookup_with_headers(_req.headers()) {
        Ok(d) => d,
        Err(_) => panic!("Error during lookup")
    };
    let body = format!("Detected device: {} - {} ", device.get_device_id(),
                       device.get_virtual_capability("complete_device_name").unwrap());
    return body;
}

examples/multithread_example.rs
use std::sync::Arc;
use std::thread;

use wurfl::{Wurfl, WurflCacheProvider};

fn main() {

    // This example application spawns 16 threads that share a single WURFL engine object and
    // perform a device detection and a get_all_device_ids, plus device calls to get_det_device_id, get_capability and get_virtual_capability

    let wurfl_path = "test_files/wurfl.zip";
    let engine = match Wurfl::new(wurfl_path, None, None,
                                  WurflCacheProvider::NoCache, None) {
        Ok(e) => e,
        Err(we) => panic!("{}", we.msg)
    };

    println!("Created engine version: {} ", engine.get_api_version());
    let safe_engine = Arc::new(engine);

    let _detection = {
        let t = thread::spawn(move || {
            let engine_ref = &safe_engine;
            for i in 1..16 {
                println!("Spawning thread: {}", i);
                let dev_ids = engine_ref.get_all_device_ids();
                println!("Got {} device IDs ", dev_ids.len());
                let ua = "Mozilla/5.0 (Linux; U; Android 8.1.0; in-id; CPH1901 Build/OPM1.171019.026) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/53.0.2785.134 Mobile Safari/537.36 OppoBrowser/15.5.1.10";
                let dev = engine_ref.lookup_useragent(ua);
                if dev.is_ok() {
                    let d = dev.unwrap();
                    println!("Got {} device ID for detected device ", d.get_device_id());
                    println!("Device brand {}  ", d.get_capability("brand_name").unwrap());
                    println!("Device is smartphone? {}  ", d.get_virtual_capability("is_smartphone").unwrap());
                    println!("---------------------------------------------------------------------");
                } else {
                    panic!("error detecting device {}", dev.err().unwrap().to_string());
                }
            };
        });
        // wait for all threads to end
        t.join().unwrap();
        println!("Ending Multi thread application - main thread ");
    };
}

examples/example.rs
use wurfl::*;

fn main() {
    println!("Starting WURFL wrapper usage sample!");
    let wurfl_path = "/usr/share/wurfl/wurfl.zip";
    let wurfl_res = Wurfl::new(wurfl_path, None, None,
                               WurflCacheProvider::LRU, Some("100000"));
    let engine = match wurfl_res {
        Ok(engine) => engine,
        Err(error) => panic!("Problem initializing wurfl: {:?}", error),
    };
    println!("WURFL API Version created: {}", engine.get_api_version());
    println!("Last load time:  {}", engine.get_last_load_time());
    println!("WURFL info:  {}", engine.get_info());

    let capability_names = engine.get_all_caps();
    // get an iterator from it and print first element
    println!("Printing capability name from Iterator");
    let mut i = capability_names.iter();
    let s = i.next();
    println!("> {} <", s.unwrap());
    // print all capabilities name via for loop (= implicit IntoIterator usage)
    println!("Printing all capability names --------------------------");
    for c in capability_names {
        println!("< {} >", c);
    }
    println!("---------------------------------------------------------");

    println!("Printing all virtual capability names -------------------");
    let vcap_names = engine.get_all_vcaps();
    for c in vcap_names {
        println!("< {} >", c);
    }
    println!("---------------------------------------------------------");

    let ua = "Mozilla/5.0 (Linux; Android 6.0.1; Redmi 4A Build/MMB29M; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/60.0.3112.116 Mobile Safari/537.36";
    let device_res2 = engine.lookup_useragent(ua);
    let d2 = match device_res2 {
        Ok(d2) => {
            println!("Device 2 ID: {}", d2.get_device_id());
            println!("Device 2 is smartphone?: {}", d2.get_virtual_capability("is_smartphone").unwrap());
            d2
        }
        Err(_) => panic!("Lookup user agent failed")
    };
    println!("Device 2 ID - outside result closure: {}", d2.get_device_id());

    // Multiple lookup results added to vector
    let uas: &[&str] = &[
        "Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 13_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.5 Mobile/15E148 Snapchat/10.77.5.59 (like Safari/604.1)",
        "Mozilla/5.0 (X11; Linux x86_64; rv:45.0) Gecko/20100101 Thunderbird/45.3.0",
        "Mozilla/5.0 (BlackBerry; U; BlackBerry 9780; en) AppleWebKit/534.8+ (KHTML, like Gecko) Version/6.0.0.480 Mobile Safari/534.8+",
        "Mozilla/5.0 (Linux; U; Android 5.0; en-US; E2303 Build/26.1.A.3.111) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/57.0.2987.108 UCBrowser/12.9.10.1159 Mobile Safari/537.36"
    ];
    let mut results = Vec::new();
    for ua in uas {
        let rd = engine.lookup_useragent(ua);
        match rd {
            Ok(d) => results.push(d),
            Err(_) => panic!("Lookup user agent failed")
        };
    }

    // print some data
    let cap_names: &[&str] = &["brand_name", "model_name"];
    println!("----- Complete device names of last 5 detections -----");
    for device in &results {
        println!("{}", device.get_virtual_capability("complete_device_name").unwrap());
        match device.get_capabilities(cap_names) {
            Ok(caps_map) => {
                for (k, v) in caps_map {
                    println!("Cap from get_capabilities <{}:{}>", k, v)
                }
            }
            Err(_) => panic!("device get_capabilities failed!")
        };
    }

    // NOTE: why am I using a reference here?
    // Because using the vector itself would move its ownership to the iterator generated under the hood in the for loop,
    // thus making impossible to use it again in the second loop and making the compiler complain
    println!("----- Check if it's a tablet for the last 5 detections -----");
    for device in &results {
        println!("Is device a tablet? {}", device.get_capability("is_tablet").unwrap());
    }
}
